doc/Language/compilation.pod6

```diff
@@ -0,0 +1,291 @@
+In other words: repositories are managed as a I<linked list>.
+The important difference to the traditional array is, that when going through the list, each object has a say in whether
+to pass along a request to the next-repo or not.
+C<Raku> sets up a standard set of repositores, i.e. the "perl", "vendor" and "site" repositories, just like you know them from C<Perl>.
+In addition, we set up a "home" repository for the current user.
+
+Repositories must implement the C<need> method.
+A C<use> or C<require> statement in C<Raku> code is basically translated to a call to C<B<$*REPO>>'s C<need> method.
+This method may in turn delegate the request to the next-repo.
+
+=begin code
+role CompUnit::Repository {
+    has CompUnit::Repository $.next is rw;
+
+    method need(CompUnit::DependencySpecification $spec,
+                CompUnit::PrecompilationRepository $precomp,
+                CompUnit::Store :@precomp-stores
+                --> CompUnit:D
+                )
+        { ... }
+    method loaded(
+                --> Iterable
+                )
+        { ... }
+
+    method id( --> Str )
+        { ... }
+}
+=end code
+
+=head2 Repositories
+
+Rakudo comes with several classes that can be used for repositories.
+The most important ones are C<CompUnit::Repository::FileSystem> and C<CompUnit::Repository::Installation>.
+The FileSystem repo is meant to be used during module development and actually works just like C<Perl> when
+looking for a module.
+It doesn't support versions or auths and simply maps the short-name to a file system path.
+
+The Installation repository is where the real smarts are. When requesting a module, you will usually either do it
+via its exact long name, or you say something along the lines of "give me a module that matches this filter".
+Such a filter is given by way of a C<CompUnit::DependencySpecification> object which has fields for
+=item short-name,
+=item auth-matcher,
+=item version-matcher and
+=item api-matcher.
+
+When looking through candidates, the Installation repository will smart match a module's long name against this
+DependencySpecification or rather the individual fields against the individual matchers.
+Thus a matcher may be some concrete value, a version range or even a regex.
+
+Loading the meta data of all installed distributions would be prohibitively slow.
+Instead, we use the file system as a kind of database.
+We store not only a distribution's files but also create indices for speeding up lookups.
+One of these indices comes in the form of directories named after the short-name of installed modules.
+However most of the file systems in common use today cannot handle Unicode names, so we cannot just use
+module names directly.
+This is where the now infamous SHA-1 hashes enter the game.
+The directory names are the ASCII encoded SHA-1 hashes of the UTF-8 encoded module short-names.
+
+In these directories we find one file per distribution that contains a module with a matching short name.
```

Comment:
This is definitely true. However, we do not yet have alternative strategies, and anyone using Raku will see these SHA-1 type names. Providing some context about what they mean will help give understanding. For instance, I got annoyed with them until I read this talk and realised what they were and why .precomp directories were scattered around my Raku directories.

Perhaps a note that this is an implementation detail and may change in future?

Reply:
I'd definitely opt for implementation detail that may change in the future, so people will test its behavior extensively in case it changes (or disappears)

Reply:
added text to reflect comments